### PR TITLE
[SDP-1145] Deactivate Tenant

### PR DESCRIPTION
**Pull request by ziyliu**
### What
- Allow admin users of `PATCH /tenants/{id}` to activate/deactivate tenant if certain conditions are met.
- Prevent `GET /tenants/...` from returning deactivated tenants.

### Why

[TODO: Why this change is being made. Include any context required to understand the why.]

### Known limitations

[TODO or N/A]

#### PR Structure

* [x] This PR has a reasonably narrow scope (if not, break it down into smaller PRs).
* [x] This PR title and description are clear enough for anyone to review it.
* [x] This PR does not mix refactoring changes with feature changes (split into two PRs otherwise).

#### Thoroughness

* [x] This PR adds tests for the new functionality or fixes.
* [x] This PR contains the link to the Jira ticket it addresses.

#### Configs and Secrets

* [x] No new CONFIG variables are required -OR- the new required ones were added to the helmchart's [`values.yaml`] file.
* [x] No new CONFIG variables are required -OR- the new required ones were added to the deployments ([`pr-preview`], [`dev`], [`demo`], `prd`).
* [x] No new SECRETS variables are required -OR- the new required ones were mentioned in the helmchart's [`values.yaml`] file.
* [x] No new SECRETS variables are required -OR- the new required ones were added to the deployments ([`pr-preview secrets`], [`dev secrets`], [`demo secrets`], `prd secrets`).

#### Release

* [ ] This is not a breaking change.
* [ ] **This is ready for production.**. If your PR is not ready for production, please consider opening additional complementary PRs using this one as the base. Only merge this into `develop` or `main` after it's ready for production!

#### Deployment

* [ ] Does the deployment work after merging?

[`values.yaml`]: ../helmchart/sdp/values.yaml
[`pr-preview`]: https://github.com/stellar/kube/blob/d3e4f5dd8aa4c13b45a31a5a937f3e98841171a7/kube001-dev/namespaces/common-previews/stellar-disbursement-platform/backend-helm-values
[`dev`]: https://github.com/stellar/kube/blob/d3e4f5dd8aa4c13b45a31a5a937f3e98841171a7/kube001-dev/namespaces/stellar-disbursement-platform/backend-helm-values
[`demo`]: https://github.com/stellar/kube/blob/d3e4f5dd8aa4c13b45a31a5a937f3e98841171a7/kube001-dev/namespaces/stellar-disbursement-platform/demo/demo-backend-helm-values
[`pr-preview secrets`]: https://github.com/stellar/kube/blob/d3e4f5dd8aa4c13b45a31a5a937f3e98841171a7/kube001-dev/namespaces/common-previews/externalsecrets-common-previews.yaml#L241-L346
[`dev secrets`]: https://github.com/stellar/kube/blob/d3e4f5dd8aa4c13b45a31a5a937f3e98841171a7/kube001-dev/namespaces/stellar-disbursement-platform/stellar-disbursement-platform-externalsecrets.yaml
[`demo secrets`]: https://github.com/stellar/kube/blob/d3e4f5dd8aa4c13b45a31a5a937f3e98841171a7/kube001-dev/namespaces/stellar-disbursement-platform/demo/demo-sdp-externalsecrets.yaml

### CODE REVIEW

File: stellar-multitenant/pkg/tenant/manager.go
Comment (marwen-abid) [inline stellar-multitenant/pkg/tenant/manager.go:90]: I like this refactor but there are a couple of adjustments needed I believe. 

I don't see any usages in the production code (non-test code) for the **queryParams** here. But the other issue is that we're not filtering out deactivated tenants by default. 

I believe we should change this and most methods in the manager to **filter out deactivated tenants by default** and create specific methods for fetching deactivated tenants **if needed** (can also use the generic `GetTenant` for special cases). 

In your next story, we will also need to change these methods to filter out deleted tenants by default. 

This is an example implementation of what I'm talking about (I didn't test this code) 
```golang
// GetTenant generic method that fetches a tenant based on queryParams
func (m *Manager) GetTenant(ctx context.Context, queryParams *QueryParams) (*Tenant, error) {
	if queryParams == nil {
		queryParams = &QueryParams{}
	}

	var t Tenant
	query, params := m.newManagerQuery(selectQuery, queryParams)
	err := m.db.GetContext(ctx, &t, query, params...)
	if err != nil {
		if errors.Is(err, sql.ErrNoRows) {
			return nil, ErrTenantDoesNotExist
		}
		return nil, fmt.Errorf("getting tenant: %w", err)
	}

	return &t, nil
}

func (m *Manager) GetTenantByID(ctx context.Context, id string) (*Tenant, error) {
	queryParams := &QueryParams{
		Filters: map[FilterKey]interface{}{
			FilterKeyID: id,
                         FilterKeyOutStatus: DeactivatedTenantStatus,
		},
	}
	
	return m.GetTenant(ctx, queryParams)
}

func (m *Manager) GetDeactivatedTenantByID(ctx context.Context, id string) (*Tenant, error) {
	queryParams := &QueryParams{
		Filters: map[FilterKey]interface{}{
			FilterKeyID: id,
			FilterKeyStatus: DeactivatedTenantStatus,
		},
	}
	
	return m.GetTenant(ctx, queryParams)
}

func (m *Manager) GetTenantByName(ctx context.Context, name string) (*Tenant, error) {
	queryParams := &QueryParams{
		Filters: map[FilterKey]interface{}{
			FilterKeyName: name,
                         FilterKeyOutStatus: DeactivatedTenantStatus,
		},
	}
	
	return m.GetTenant(ctx, queryParams)
}

... etc. 
```

File: stellar-multitenant/pkg/internal/services/status_change_validator.go
Comment (ziyliu) [reply stellar-multitenant/pkg/internal/services/status_change_validator.go:22]: if a tenant is `deactivated` and the requested status is `activated`, this validator function will avoid returning an error and the calling handler proceeds w/ the rest of the update.

File: stellar-multitenant/pkg/tenant/manager.go
Comment (marcelosalloum) [inline stellar-multitenant/pkg/tenant/manager.go:63]: Can you wrap this err plz?

File: stellar-multitenant/pkg/tenant/manager.go
Comment (marcelosalloum) [inline stellar-multitenant/pkg/tenant/manager.go:80]: Can you add an optional `filters` parameter, so callers can request the tenants with custom (or without any) filters?
* If they send nil, you replace by the default one already in place
* If they send anything else, you use whatever they sent, including `map[FilterKey]interface{}{}`

File: stellar-multitenant/pkg/tenant/manager.go
Comment (marcelosalloum) [inline stellar-multitenant/pkg/tenant/manager.go:95]:
```suggestion
// GetTenant is a generic method that fetches a tenant based on queryParams.
```

**Review by marcelosalloum — COMMENTED** [review]
I'm still finishing the review, but here are some small feedback:

File: stellar-multitenant/internal/services/status_change_validator.go
Comment (marwen-abid) [inline stellar-multitenant/internal/services/status_change_validator.go:43]: I see that we avoid deactivating an instance that has `NonTerminalStatus`. 
`DraftPaymentStatus, ReadyPaymentStatus, PendingPaymentStatus, PausedPaymentStatus`

I think it's ok to deactivate an instance with `DraftPaymentStatus` and potentially even `PausedPaymentStatus`. Thoughts on this @marcelosalloum ?

**Review by marwen-abid — COMMENTED** [review]
Thank you for addressing my concerns. The changes look good to me. 
I'll let @marcelosalloum finish his review.

File: stellar-multitenant/internal/services/status_change_validator.go
Comment (marcelosalloum) [reply stellar-multitenant/internal/services/status_change_validator.go:43]: I agree! @ziyliu: I think you can replace the method `data.PaymentNonTerminalStatuses()` with something that denotes "active/inProgress/pending payments".

File: cmd/serve_test.go
Comment (marcelosalloum) [inline cmd/serve_test.go:209]: 💡 According to the Go naming conventions, acronyms should be either all-uppercased or all-lowercased. Plz, refer to the initialisms section from Go's official [CodeReviewComments wiki](https://go.dev/wiki/CodeReviewComments).

File: stellar-multitenant/internal/httphandler/tenants_handler_test.go
Comment (marcelosalloum) [inline stellar-multitenant/internal/httphandler/tenants_handler_test.go:669]: 💡  nit: how about rename `"unsuccessfully updates..."` to `"cannot update..."`, like you did in line 687?

File: stellar-multitenant/internal/httphandler/tenants_handler_test.go
Comment (marcelosalloum) [inline stellar-multitenant/internal/httphandler/tenants_handler_test.go:169]: 💡  Can you be more explicit in the variable name? `dTnt` will probably not make sense when someone look at this code months from now.

File: stellar-multitenant/internal/httphandler/tenants_handler_test.go
Comment (marcelosalloum) [inline stellar-multitenant/internal/httphandler/tenants_handler_test.go:72]: 🤔 shouldn't you be closing it after the `readAll` call?

File: stellar-multitenant/internal/services/status_change_validator_test.go
Comment (marcelosalloum) [inline stellar-multitenant/internal/services/status_change_validator_test.go:158]: 💡  For better debuggability and improved test isolation, it'd be better to declare and assert the expectations of the mock inside t.Run. Otherwise, one failed subtest can fail all the subsequent ones.

File: stellar-multitenant/pkg/serve/serve_test.go
Comment (marcelosalloum) [inline stellar-multitenant/pkg/serve/serve_test.go:57]: Same here about achronyms: Mtn... -> MTN...

File: stellar-multitenant/pkg/tenant/manager.go
Comment (marcelosalloum) [inline stellar-multitenant/pkg/tenant/manager.go:176]: 👋 🤔  What happens if you try to deactivate a default tenant?

File: stellar-multitenant/pkg/tenant/manager.go
Comment (marcelosalloum) [inline stellar-multitenant/pkg/tenant/manager.go:109]: 👋 💡 Can't you improve the core reusability by calling GetTenant from GetTenantByID, GetTenantByName, GetTenantByIDOrName, and GetDefault?

**Review by marcelosalloum — COMMENTED** [review]
Thank you so much for working on this.

I've left a few comments, the most important ones I marked with 👋 .

👋  Also, could you plz include a test for `ResolveTenantFromRequestMiddleware` to ensure that deactivated tenants are ignored on that middleware? This is the most sensitive place where tenants are used, I believe, so it'd be nice to add a safeguard there.

File: cmd/serve_test.go
Comment (ziyliu) [reply cmd/serve_test.go:209]: Thanks, I was following the precedent set [here](https://github.com/stellar/stellar-disbursement-platform-backend/blob/develop/internal/serve/serve.go#L58), but I suppose we can fix that another time.

File: stellar-multitenant/internal/services/status_change_validator_test.go
Comment (ziyliu) [reply stellar-multitenant/internal/services/status_change_validator_test.go:158]: got itt!

File: stellar-multitenant/internal/httphandler/tenants_handler_test.go
Comment (ziyliu) [reply stellar-multitenant/internal/httphandler/tenants_handler_test.go:72]: I think this is the preferred standard since it comes immediately after the resp is opened? for [example](https://github.com/stellar/stellar-disbursement-platform-backend/blob/eb3903e19e364d759c52da5c61cce07d2070bf93/internal/serve/validators/recaptcha.go#L52-L58)

File: stellar-multitenant/pkg/tenant/manager.go
Comment (ziyliu) [reply stellar-multitenant/pkg/tenant/manager.go:109]: done! I don't think this can be applicable to `GetDefault` though

File: stellar-multitenant/pkg/tenant/manager.go
Comment (ziyliu) [reply stellar-multitenant/pkg/tenant/manager.go]: addressed in [993ea99](https://github.com/stellar/stellar-disbursement-platform-backend/pull/253/commits/993ea99d2a28bdae1e02e8c03d9b44abc85cb384)

File: stellar-multitenant/internal/httphandler/tenants_handler.go
Comment (marcelosalloum) [inline stellar-multitenant/internal/httphandler/tenants_handler.go:35]: I think the `GET {adminServer}/tenants` should return deactivated tenants too, so the Host is able to see the IDs of tenants they want to deactivate. WDYT @ziyliu?

Would this change be enough for that?
```suggestion
	tnts, err := t.Manager.GetAllTenants(ctx, &tenant.QueryParams{})
```

**Review by marcelosalloum — APPROVED** [review]
LGTM! Plz feel free to merge it after addressing the comments below, as well as the mutual exclusion of "default tenants" vs "deactivated tenants".

File: stellar-multitenant/internal/httphandler/tenants_handler.go
Comment (ziyliu) [reply stellar-multitenant/internal/httphandler/tenants_handler.go:35]: Makes sense to me, and @marwen-abid also agrees.

**Pull request merged**
